### PR TITLE
Change the way TDD core skill is assessed

### PR DESCRIPTION
**Pull request by adrianclay**
### Major changes

* Triangulation is now in Giraffe
* Acceptance criteria added to Dragon
* Removed a number of soggy requirements from Dragon
* Make Giraffe language agnostic
* Reduce the cost of adding additional languages to the Wolf and Dragon levels
* Added the concept of a _secret_ marking scheme

### TODO

* Create the secret marking scheme
* Create a collection of test doubles to be used and tested against
* Is Chirp app still suitable for Dragon?

### CODE REVIEW

**Comment by danielburnley** [comment]
I like the change that's been made in regards to making it language agnostic, I think that should definitely help reduce the monotony/rails specific-ness of the assessment and probably make it a lot smoother as a result. Having spent half of mine asking questions about rails itself as opposed to TDD, i look forward to seeing what happens going forward

File: core-skills/tdd/README.md [hunk 1]
```diff
@@ -24,91 +24,104 @@
...
-- [ ] Can describe expected behaviours from the perspective of the user.
+- Can describe expected behaviours from the perspective of the user.
   - Required - Can write test names which describe the behaviour of the software from a user’s perspective.
 
+- Can describe the situations where using triangulation helps improve test quality.
+  - Required - Can explain what the potential flaws of not including triangulation are.
+  - Required - Can explain how triangulation overcomes these flaws.
+
+- Can demonstrate triangulation when TDDing an example piece of code.
+  - Required - Production code cannot be easily slimed, due to the use of triangulation.
```
Comment (craigjbass) [inline core-skills/tdd/README.md]: This reads weirdly.

Comment (craigjbass) [reply core-skills/tdd/README.md]: `Production code cannot be easily slimed` - should probs link to the practice of generalising code as tests become more specific

File: core-skills/tdd/README.md [hunk 2]
```diff
@@ -24,91 +24,104 @@
...
-  - Required - No example of this, unless explained with good reason
-
-- [ ] Can identify smells within code they have written
-
-- [ ] Can identify smells within somebody else's code
+Review and refactor a preprepared test suite to improve performance.
 
-- [ ] Can explain the balancing act of achieving conciseness of a test suite
+- Can write an acceptance test around an existing feature.
+  - Required - Write an acceptance test around a feature of an existing application which verifies the feature works as intended.
```
Comment (craigjbass) [inline core-skills/tdd/README.md]: How come this is for an existing application. Implies not-test first?

We haven't defined what an acceptance test is... Is it [subcutaneous](https://martinfowler.com/bliki/SubcutaneousTest.html) or something more high-level (e.g. through the UI).

I always use Acceptance Test to refer to subcutaneous tests, and Feature Test to refer to through the UI testing.

File: core-skills/tdd/README.md [hunk 3]
```diff
@@ -24,91 +24,104 @@
...
+- Can identify smells within somebody else's code.
+  - Required - Can list three valid smells per the invigilator's guide.
 
-- [ ] Can identify structural defects that result in poor test suite performance
+- Can describe why test suite performance is important.
+  - Required - See invigilator's guide.
 
-- [ ] Can identify that a test suite has poor test suite performance
+- Can identify causes of poor performance in a test suite.
+  - Required - Identify three of the secret causes of poor performance.
```
Comment (craigjbass) [inline core-skills/tdd/README.md]: three secret causes :O

**Comment by craigjbass** [comment]
Have we updated the accompanying sections to reflect the new marking guide?

File: core-skills/tdd/README.md (hunk 1 above)
Comment (zuren) [reply core-skills/tdd/README.md]: Please suggest a change.

File: core-skills/tdd/README.md (hunk 2 above)
Comment (zuren) [reply core-skills/tdd/README.md]: Agreed, it's not test first in this case. We felt that being able to demonstrate writing an acceptance test (discussed as either subcutaneous or through the UI) around an existing feature would demonstrate enough of the skills, and be a good first step in TDDing legacy code. Which is basically Dragon.

**Comment by zuren** [comment]
They should all be updated @craigjbass - did you notice any that weren't?

File: core-skills/tdd/README.md (hunk 3 above)
Comment (zuren) [reply core-skills/tdd/README.md]: Better than "knows the secret smells"

Comment (zuren) [reply core-skills/tdd/README.md]: I'm not a big fan of the secrecy thing. I feel it's complicated and brittle.

We felt there were particular bits of knowledge that _must_ be known by the candidate, and that putting them in the marking guide wouldn't test that.

File: core-skills/tdd/README.md (hunk 1 above)
Comment (craigjbass) [reply core-skills/tdd/README.md]: `- Required - Production code is generalised only when there are failing tests that require generalised production code.`

File: core-skills/tdd/README.md (hunk 2 above)
Comment (craigjbass) [reply core-skills/tdd/README.md]: I don't like that we have a pun on what "Acceptance Test" means at Made Tech. 

Recognising that we're talking about Subcutaneous Testing vs UI testing is an _important_ distinction.

**Comment by zuren** [comment]
@craigjbass I'll amend Dragon in a separate PR, I want to get this merged for assessments running now.

**Pull request merged**